test/micro/org/openjdk/bench/jdk/incubator/vector/StoreMaskTrueCount.java

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) 2023, Arm Limited. All rights reserved.
+// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+//
+// This code is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License version 2 only, as
+// published by the Free Software Foundation.
+//
+// This code is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// version 2 for more details (a copy is included in the LICENSE file that
+// accompanied this code).
+//
+// You should have received a copy of the GNU General Public License version
+// 2 along with this work; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+// Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+// or visit www.oracle.com if you need additional information or have any
+// questions.
+//
+//
+package org.openjdk.bench.jdk.incubator.vector;
+
+import java.util.concurrent.TimeUnit;
+import java.util.Random;
+import jdk.incubator.vector.*;
+import org.openjdk.jmh.annotations.*;
+
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@State(Scope.Thread)
+@Fork(jvmArgsPrepend = {"--add-modules=jdk.incubator.vector"})
+public class StoreMaskTrueCount {
+    private static final VectorSpecies<Short> S_SPECIES = ShortVector.SPECIES_PREFERRED;
+    private static final VectorSpecies<Integer> I_SPECIES = IntVector.SPECIES_PREFERRED;
+    private static final VectorSpecies<Long> L_SPECIES = LongVector.SPECIES_PREFERRED;
+    private static final int LENGTH = 1024;
+    private static final Random RD = new Random();
+    private static boolean[] ba;
+
+    static {
+        ba = new boolean[LENGTH];
+        for (int i = 0; i < LENGTH; i++) {
+            ba[i] = RD.nextBoolean();
+        }
+    }
+
+    @Benchmark
+    public static int testShort() {
+        VectorMask<Short> m = VectorMask.fromArray(S_SPECIES, ba, 0);
+        int res = 0;
+        for (int i = 0; i < LENGTH; i++) {
+            m = m.not();
+            res += m.trueCount();
+        }
+
+        return res;
+    }
+
+    @Benchmark
+    public static int testInt() {
+        VectorMask<Integer> m = VectorMask.fromArray(I_SPECIES, ba, 0);
+        int res = 0;
+        for (int i = 0; i < LENGTH; i++) {
+            m = m.not();
+            res += m.trueCount();
+        }
+
+        return res;
+    }
+
+    @Benchmark
+    public static int testLong() {
+        VectorMask<Long> m = VectorMask.fromArray(L_SPECIES, ba, 0);
+        int res = 0;
+        for (int i = 0; i < LENGTH; i++) {
+            m = m.not();
+            res += m.trueCount();
+        }
```

Comment:
This looks like it might be removed by loop opts. I think you might need a blackhole somewhere.

Reply:
> This looks like it might be removed by loop opts. I think you might need a blackhole somewhere.

```m``` will be updated in every iteration of this loop, so ```m``` is not a loop-invariants actually.
I can see the assembly code of this loop by using JMH perfasm.

Reply:
> > This looks like it might be removed by loop opts. I think you might need a blackhole somewhere.
> 
> `m` will be updated in every iteration of this loop, so `m` is not a loop-invariants actually. I can see the assembly code of this loop by using JMH perfasm.

Isn't it? Looks to me like all it does is flip `m` each time. Whether or not this code is optimized today isn't relevant.

So it's the same as

```
        for (int i = 0; i < LENGTH/2; i++) {
            res += m.trueCount();
        }
        m = m.not();
        for (int i = 0; i < LENGTH/2; i++) {
            res += m.trueCount();
        } 
```

... which is trivially optimizable, no?

Reply:
> > > This looks like it might be removed by loop opts. I think you might need a blackhole somewhere.
> > 
> > 
> > `m` will be updated in every iteration of this loop, so `m` is not a loop-invariants actually. I can see the assembly code of this loop by using JMH perfasm.
> 
> Isn't it? Looks to me like all it does is flip `m` each time. Whether or not this code is optimized today isn't relevant.
> 
> So it's the same as
> 
> ```
>         for (int i = 0; i < LENGTH/2; i++) {
>             res += m.trueCount();
>         }
>         m = m.not();
>         for (int i = 0; i < LENGTH/2; i++) {
>             res += m.trueCount();
>         } 
> ```
> 
> ... which is trivially optimizable, no?

Sorry for the delay.

Yes, actually they do the same thing, though current C2 compiler cannot do such optimization so far. Anyway, I have updated this benchmark to avoid potential optimization and ensure that we can measure performance effectively.